extensions-core/lookups-cached-single/src/main/java/org/apache/druid/server/lookup/LoadingLookup.java

```diff
@@ -74,14 +75,22 @@ public String apply(@Nullable final String key)
     }
 
     final String presentVal;
-    try {
-      presentVal = loadingCache.get(keyEquivalent, new ApplyCallable(keyEquivalent));
+    presentVal = this.loadingCache.getIfPresent(keyEquivalent);
+    if (presentVal != null) {
       return NullHandling.emptyToNullIfNeeded(presentVal);
     }
-    catch (ExecutionException e) {
-      LOGGER.debug("value not found for key [%s]", key);
+
+    String val = this.dataFetcher.fetch(keyEquivalent);
+    if (val == null) {
       return null;
     }
+
+    Map<String, String> map = new HashMap<>();
+    map.put(keyEquivalent, val);
+
+    this.loadingCache.putAll(map);
+
+    return NullHandling.emptyToNullIfNeeded(val);
```

Comment:
The new change suggests that if we don't find the value in the cache we get it from the dataFetcher (DB or URI) and then we update the cache, so in the next probable query we have the value.
(The old code, if I get right, was just getting data from the loading cache and if it just returns the value from there whether it exists or not, the cache is never updated).

Reply:
Also, I think some of the changes in the test layer could also be reverted with this?

Reply:
Have you tested the case where you have data in Druid that has no mapping key in the DB? (If I remember well that was one of the issues I encountered and fixed by this part of the code).
i.e. we have a username with ID 1  in Druid but in DB we have only a username with ID 2.

Reply:
@TessaIO Can you please elaborate? Do you mean if the lookup is configured with a JDBC source that has only `id1->val1`, then what happens if we try to look for `id2` in the lookup?

Reply:
Let's suppose that we have 2 rows in Druid, ID 1, and ID 2, and in the JDBC source we have only 1 row with ID 1. When we execute a query (Groupby for e.g) with a lookup in the aggregation, it will probably fail (according to my testing) with a null pointer exception because ID 2 doesn't exist in the JDBC source (according to the old code of course).

Reply:
Tl;dr: The approach in this latest patch looks okay to me.

1. If we use `getIfPresent()` and not the `ApplyCallable` approach, we should remove the unused `ApplyCallable` class
2. Add a test to ensure fetching a non-existent key from the data fetcher works as expected  - does the `testApplyWithNullCase()` case already cover that?

-----

**Alternative idea:**
@Akshat-Jain, re this:
> While the above query does work with your code changes, we would lose out on the "collection of cache statistics" part which was happening in the loadingCache.get() flow.

Are you referring to the off heap/on heap stats? I see for the off heap loading cache, the stats are in fact only modified in the `getIfPresent()` [call](https://github.com/apache/druid/blob/master/extensions-core/lookups-cached-single/src/main/java/org/apache/druid/server/lookup/cache/loading/OffHeapLoadingCache.java#L113).  Not sure why those stats are not adjusted in `get()` call, which seems like a separate bug to me.


A similar approach to what Akshat is referring to could also work by throwing an exception. Fwiw, this is called out in the [javadocs](https://github.com/apache/druid/blob/master/extensions-core/lookups-cached-single/src/main/java/org/apache/druid/server/lookup/cache/loading/LoadingCache.java#L74) for `LoadingCache#get(K, Callable)`:

```
    * <p><b>Warning:</b> as with {@link CacheLoader#load}, {@code valueLoader} <b>must not</b> return
    * {@code null}; it may either return a non-null value or throw an exception.
```
So it looks throwing an exception is another way to handle this if we go down the `get(K, Callable)`. So we can perhaps change `ApplyCallable#call()` to explicitly check for null value and throw an exception, but it must be caught and handled in `apply()` to return null. I don't necessarily like this approach because it relies on throwing and handling checked exceptions just to work around the underlying library's implementation.

Reply:
> Are you referring to the off heap/on heap stats

@abhishekrb19 I was referring to the on-heap cache stats. `loadingCache.get()` ends up hitting `LoadingCache#getAndRecordStats`, which doesn't get hit in the PR changes.

Reply:
@Akshat-Jain @abhishekrb19 Would it be okay if we keep the implementation as is?